sycl/include/sycl/types.hpp

```diff
@@ -2087,298 +2090,120 @@ __SYCL_RELLOGOP(||)
+      "or 16 are supported");
+  static_assert(!std::is_same_v<V, void>, "Incorrect data type for sycl::vec");
+};
 
-#if (!defined(_HAS_STD_BYTE) || _HAS_STD_BYTE != 0)
-#define __SYCL_DECLARE_SCALAR_BYTE_CONVERTER                                   \
-  template <> class BaseCLTypeConverter<std::byte, 1> {                        \
-  public:                                                                      \
-    using DataType = __SYCL_GET_SCALAR_CL_TYPE(uchar);                         \
+#ifdef __SYCL_USE_EXT_VECTOR_TYPE__
+template <typename T, int N> struct VecStorageImpl {
+  using DataType = T __attribute__((ext_vector_type(N)));
+};
+#else
+// When ext_vector_type is not available, we rely on cl_* types from CL/cl.h
+// to represent vec storage.
+template <typename T, int N> struct VecStorageImpl;
+#define __SYCL_DEFINE_VECSTORAGE_IMPL(type, cl_type, num)                      \
+  template <> struct VecStorageImpl<type, num> {                               \
+    using DataType = ::cl_##cl_type##num;                                      \
   };
-#endif
-#define __SYCL_DECLARE_SCALAR_CONVERTER(base)                                  \
-  template <> class BaseCLTypeConverter<base, 1> {                             \
-  public:                                                                      \
-    using DataType = __SYCL_GET_SCALAR_CL_TYPE(base);                          \
+#define __SYCL_DEFINE_VECSTORAGE_IMPL_FOR_TYPE(type, cl_type)                  \
+  __SYCL_DEFINE_VECSTORAGE_IMPL(type, cl_type, 2)                              \
+  __SYCL_DEFINE_VECSTORAGE_IMPL(type, cl_type, 3)                              \
+  __SYCL_DEFINE_VECSTORAGE_IMPL(type, cl_type, 4)                              \
+  __SYCL_DEFINE_VECSTORAGE_IMPL(type, cl_type, 8)                              \
+  __SYCL_DEFINE_VECSTORAGE_IMPL(type, cl_type, 16)
+
+__SYCL_DEFINE_VECSTORAGE_IMPL_FOR_TYPE(std::int8_t, char)
+__SYCL_DEFINE_VECSTORAGE_IMPL_FOR_TYPE(std::int16_t, short)
+__SYCL_DEFINE_VECSTORAGE_IMPL_FOR_TYPE(std::int32_t, int)
+__SYCL_DEFINE_VECSTORAGE_IMPL_FOR_TYPE(std::int64_t, long)
+__SYCL_DEFINE_VECSTORAGE_IMPL_FOR_TYPE(std::uint8_t, uchar)
+__SYCL_DEFINE_VECSTORAGE_IMPL_FOR_TYPE(std::uint16_t, ushort)
+__SYCL_DEFINE_VECSTORAGE_IMPL_FOR_TYPE(std::uint32_t, uint)
+__SYCL_DEFINE_VECSTORAGE_IMPL_FOR_TYPE(std::uint64_t, ulong)
+__SYCL_DEFINE_VECSTORAGE_IMPL_FOR_TYPE(float, float)
+__SYCL_DEFINE_VECSTORAGE_IMPL_FOR_TYPE(double, double)
+#undef __SYCL_DEFINE_VECSTORAGE_IMPL_FOR_TYPE
+#undef __SYCL_DEFINE_VECSTORAGE_IMPL
+#endif // __SYCL_USE_EXT_VECTOR_TYPE__
+// select_apply_cl_t selects from T8/T16/T32/T64 basing on
+// sizeof(IN).  expected to handle scalar types in IN.
+template <typename IN, typename T8, typename T16, typename T32, typename T64>
+using select_apply_cl_t =
+    conditional_t<sizeof(IN) == 1, T8,
+                  conditional_t<sizeof(IN) == 2, T16,
+                                conditional_t<sizeof(IN) == 4, T32, T64>>>;
```

Comment:
Note for myself and reviewers: first template argument was called `T` instead of `IN` on purpose: `windows.h` defines macro `IN` which screws up this code